### ISSUE TITLE
Callback chain broken in unexpected way with combination of association + callbacks

### ISSUE THREAD

**Issue by eileencodes** [rails/rails#36132]
This is related to #36082 in that the issue we're seeing in the GitHub application is caused by the same commit but I think we have discovered some buggy behavior.

In our application we have some models that have a corresponding "archived" table so we can "delete" records and their associations without having to actually delete them. We do a replace into the archived table and then delete from the original table. There are reasons we do this, and I could get into why, but I can demonstrate an unexpected change in behavior in Rails due to #36082 in the callback chain without explaining why we do that in our app.

In our case we have models that have `after_commit on: :destroy` callbacks that we use to loop through and throttle the deletion of records. For large repositories this is really important for us.

Since updating Rails last week we've had a few test failures we're working through, one is showing our `after_commit` callbacks are **not** firing at all on destroy for our archive code. I've reproduced the issue outside of our app in the script below. In order to reproduce the issue a few things must be true:

1) A model (Post) has a `has_many` association (Comment) that is `dependent: :destroy` and a lambda adding an order (I don't know why but the lambda is required to repro this, it could be empty, or where or whatever but the lambda is one part of reproducing this).
2) The `has_many` model (Comment) belongs to Post, and has an `after_destroy` callback that updates the Post record. In our app we use this to update the number of downloads before a repo is deleted. A `before_destroy` callback also breaks the chain if it also updates the Post.

With this scenario any callbacks that are `after_commit on: :destroy` on the Post model will NOT be called and will be completely skipped. If I change this to an `after_destroy` it works fine.

Our code may not be the best pattern here but my main concern is how hard it was to figure out what broke and am wary that this will break other apps in unexpected ways.

I've created a reproduction script that demonstrates the issue but doesn't represent a real world case. The script shows how the `after_commit` callback is never fired by using an array to capture that it was fired.

This test passes before https://github.com/rails/rails/pull/35920 was introduced and fails afterwards. Here's the script

```ruby
# frozen_string_literal: true

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t| 
    t.string :title
  end 

  create_table :comments, force: true do |t| 
    t.text :comment_body
    t.integer :post_id
  end 
end

class Post < ActiveRecord::Base
  # This never gets called when order and an after_destroy
  # callback that updates Post is called
  #
  # This also works if we use `after_destroy`. I think what happens here is that
  # when we made after_commit not fire for if a record isn’t saved Rails is
  # considering this record not saved.
  after_commit on: :destroy do
    self.callback_called << "yes"
  end

  # if we don’t have a lamdba OR don’t have a callback on comment this passes
  has_many :comments, -> { order("comments.id DESC") }, dependent: :destroy

  def callback_called
    @callback_called ||= []
  end

  def callback_called=(callback_called)
    @callback_called = [callback_called]
  end
end

class Comment < ActiveRecord::Base
  belongs_to :post

  # Remove this and Post’s after_commit callbacks will be called.
  after_destroy :save_post

  # This must update/save a comment to break the callback chain
  def save_post
    post.save!
  end
end

class BugTest < Minitest::Test
  def test_after_commit
    post = Post.create!(title: "I am a post")
    comment = post.comments.create(comment_body: "i am a comment")

    assert_equal 1, post.comments.count

    post.destroy

    # pass - post is successfully deleted
    assert !Post.exists?(post.id)
    # pass - comments are successfully deleted
    assert_equal 0, Comment.where(post_id: post.id).count
    # fail - callbacks are not called when order, dependency, and an after_destroy
    # callback is used.
    assert_equal ["yes"], post.callback_called
  end
end
```

@kamipo I'd like you to weigh in on whether this behavior is expected or not with the change you made to `after_commit`

cc/ @jhawthorn @tenderlove @cpruitt for visibility.

**Comment by eileencodes** [comment]
@matthewd asked if this was related to how in tests the records aren't truly committed but I verified in a demo application that this occurs outside of the test framework so it's not related to that.

**Comment by kamipo** [comment]
Identified. This regression is due to changing record registration order to the transaction.

In the reproduction script, `comment.post` loads new post (since automatic inverse_of is disabled due to association having lambda), then the records in the transaction will be `[<loaded (saved) post: 1>, <destroyed post: 1>]`, and `<loaded (saved) post: 1>` wins by `records.uniq`.

We should keep the registration order to avoid any regression.
I'll create a PR to fix that.

**Comment by cpruitt** [comment]
I believe we might have a similar issue related to associated records which attempt to delete their `belongs_to` association. In this modified version of Eileen's script, `Comment` will delete `Post` if there are no other comments remaining.

```ruby
# frozen_string_literal: true

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
    t.string :title
  end

  create_table :comments, force: true do |t|
    t.text :comment_body
    t.integer :post_id
  end
end

class Post < ActiveRecord::Base
  # This never gets called when order and an after_destroy
  # callback that updates Post is called
  after_commit on: :destroy do
    self.callback_called << "yes"
  end

  # If we don’t have a lamdba OR don’t have a callback on comment this passes
  has_many :comments, -> { order("comments.id DESC") }, dependent: :destroy

  def callback_called
    @callback_called ||= []
  end

  def callback_called=(callback_called)
    @callback_called = [callback_called]
  end
end

class Comment < ActiveRecord::Base
  belongs_to :post

  # Remove this and Post’s after_commit callbacks will be called.
  after_destroy :destroy_post_if_last_comment

  def destroy_post_if_last_comment
    post.destroy if post && post.comments.empty?
  end
end

class BugTest < Minitest::Test
  def test_after_commit
    post = Post.create!(title: "I am a post")
    comment = post.comments.create(comment_body: "i am a comment")

    assert_equal 1, post.comments.count

    post.destroy

    # pass - post is successfully deleted
    assert !Post.exists?(post.id)
    # pass - comments are successfully deleted
    assert_equal 0, Comment.where(post_id: post.id).count
    # fail - callbacks are not called when order, dependency, and an after_destroy
    # callback is used.
    assert_equal ["yes"], post.callback_called
  end
end
```

I don't have a whole lot of an idea about what's happening, but I did notice that multiple instances of the Post record are added to the transaction records and saw this in `ActiveRecord::Persistence#destroy`:

```
@_trigger_destroy_callback = if persisted?
  destroy_row > 0
else
  true
end
```

It seems as though the original Post record is registered, then a second Post instance for the same `id` is created via the `Comment` after destroy callback. The second `Post` instance is destroyed first. When the first `Post` instance later calls `destroy_row` there is no database record to destroy so `@_trigger_destroy_callback` is set to false. Because this first registered record is the one we check for callbacks, and `@_trigger_destroy_callback` on that record is not true the callbacks do not run.

cc/ @eileencodes @kamipo for guidance
cc/  @jhawthorn @tenderlove for visibility

**Comment by kamipo** [comment]
@cpruitt Is that a regression? It doesn't pass even if 5.2.3.

**Comment by cpruitt** [comment]
@kamipo I posted that updated test script prematurely. We have some issues at GitHub with some tests that follow a simmilar destroy callbacks pattern, but with more complexity, and show the same behavior behavior as the reproduction script. Our issue occurs with our latest pull of Rails `6-0-stable` but not 5.2.3. My attempt to reproduce it in a more simplified way wasn't very accurate I'm afraid. We're getting a better understanding of our failures and will share some more accurate details if it's relevant to this issue. I'm sorry for the unnecessary ping.

**Comment by kamipo** [comment]
I think perhaps #36190 might fix any issue at GitHub.
Let me know if still have any failure after #36190.

**Comment by eileencodes** [comment]
Hey @kamipo that fixed the remaining bugs, thank you so much!

**Comment by jeffblake** [comment]
Hi @kamipo - I'm seeing a regression here in Rails 6 with `after_rollback` callbacks not being called on new records. The callback isn't being called because of `trigger_transactional_callbacks?` returning false as the record is not persisted or destroyed.
https://github.com/rails/rails/pull/36190/files#diff-174733f2db65ef1bc53e3222d6ac0e61R381

In 5.2, the callbacks ran. I'm relying on this callback to undo some state in Redis (reserving inventory in an e-commerce app). Any thoughts?

**Comment by kamipo** [comment]
Can you create new issue with reproduction script?

I see changed behavior that `after_rollback` callbacks not being called when `update` and `destroy` are not succeeded, but on new records (i.e. `create` is not succeeded), it looks like not changed.

**Comment by jeffblake** [comment]
Yes, opened an issue here https://github.com/rails/rails/issues/36965
